src/modm/driver/display/st7586s_impl.hpp

```diff
@@ -0,0 +1,119 @@
+	DC::set();  // data mode
+	if (len > 0) {
+		SPI::transferBlocking(reinterpret_cast<const uint8_t*>(data), nullptr, len);
+	}
+	CS::set();
+	// exit with data mode on
+}
+
+template <typename SPI, typename CS, typename RST, typename DC, int Width, int Height>
+void
+St7586s<SPI, CS, RST, DC, Width, Height>::initialize()
+{
+	namespace payload = impl::st7586s::payload;
+
+	// Configure GPIO
+	RST::reset();
+	RST::setOutput();
+	CS::set();
+	CS::setOutput();
+	DC::setOutput();
+
+	// Reset display
+	modm::delay_us(10);
+	RST::set();
+	modm::delay_ms(120);
+
+	// Power ON operation flow (see datasheet)
+	sendCommand(Command::SleepOff);
+	sendCommand(Command::DisplayOff);
+	modm::delay_ms(50);  // t_{ON-V2}
+	sendCommand(Command::SetVop, payload::SetVop{0xF8, 0});  // TODO: 13.52 constexpr conversion
+	sendCommand(Command::SetBias, payload::SetBias::Ratio_1_11);
+	sendCommand(Command::SetBooster, payload::SetBooster::x8);
+	sendCommand(Command::AnalogControl, payload::AnalogControl::Enable);
+	sendCommand(Command::NLineInversion, payload::NLineInversion::FrameInversion);
+	sendCommand(Command::DisplayModeMono);  // TODO: grayscale
+	sendCommand(Command::EnableDdram, payload::EnableDdram::Enable);
+	sendCommand(Command::ScanDirection, payload::DisplayControl::ComInc
+		/* TODO: | payload::DisplayControl::SegInc */);
+	sendCommand(Command::DisplayDuty, uint8_t(Height - 1));
+	sendCommand(Command::InverseOff);
+	setWindow(0, 0, Width, Height);
+	// TODO: either clean the display or not turn it on
+	sendCommand(Command::DisplayOn);
+}
+
+template <typename SPI, typename CS, typename RST, typename DC, int Width, int Height>
+void
+St7586s<SPI, CS, RST, DC, Width, Height>::setWindow(uint16_t x, uint16_t y, uint16_t w, uint16_t h)
+{
+	namespace payload = impl::st7586s::payload;
+
+	modm_assert(x < Width, "st4586s.sw.x", "x >= Width", x);
+	modm_assert(x + w <= Width, "st4586s.sw.xw", "x + w >= Width", x + w);
+	modm_assert(y < Height, "st4586s.sw.y", "y >= Height", y);
+	modm_assert(y + h <= Height, "st4586s.sw.yh", "y + h >= Height", y + h);
+	modm_assert((x % pixelsPerByte) == 0, "st4586s.sw.x%", "x is not a multiply of PBB", x);
+	modm_assert((w % pixelsPerByte) == 0, "st4586s.sw.w%", "w is not a multiply of PBB", w);
+
+	const payload::SetColumnRow columnRange(x / pixelsPerByte, (x + w) / pixelsPerByte - 1);
+	sendCommand(Command::SetColumn, columnRange);
+
+	const payload::SetColumnRow rowRange(y, y + h - 1);
+	sendCommand(Command::SetRow, rowRange);
+}
+
+template <typename SPI, typename CS, typename RST, typename DC, int Width, int Height>
+void
+St7586s<SPI, CS, RST, DC, Width, Height>::update()
+{
+	sendCommand(Command::WriteDisplayData);
+	CS::reset();
+	// TODO: support windows other than full screen
+	// TODO: transfer the whole memory area, not individual pixels
+	for (uint16_t y = 0; y < Height; y++) {
+		for (uint16_t x = 0; x + 2 < Width; x += 3) {
+			uint8_t cell = 0;
+
+			// TODO: getting pixels this way is not very speedy
```

Comment:
@TomSaw look at this encoding. 4-level gray scale, but encoded at 3 pixels per byte.

Check page 18 of datasheet: https://www.topwaydisplay.com/sites/default/files/2020-04/ST7586S.pdf

Reply:
Is included in #665 . You can choose any of 1 - 64 bpp for your Grayscale-type

Reply:
Means more options on instantiating the `Buffer<...>`. Do you know other Displays out there with such a wild mapping @twasilczyk  ?

Reply:
Displays yes (it’s Smart Response XE and Lego Mindstorms EV3), but they are both using the same controller. I’m fine doing buffer implementation locally.